src/nimblepkg/nimscriptwrapper.nim

```diff
@@ -0,0 +1,131 @@
+# Copyright (C) Andreas Rumpf. All rights reserved.
+# BSD License. Look at license.txt for more info.
+
+## Implements the new configuration system for Nimble. Uses Nim as a
+## scripting language.
+
+import common, version, options, packageinfo, cli
+import hashes, json, os, streams, strutils, strtabs, tables, times, osproc, sets, pegs
+
+type
+  Flags = TableRef[string, seq[string]]
+  ExecutionResult*[T] = object
+    success*: bool
+    command*: string
+    arguments*: seq[string]
+    flags*: Flags
+    retVal*: T
+
+const
+  internalCmd = "e"
+  nimscriptApi = staticRead("nimscriptapi.nim")
+  nimscriptHash = $nimscriptApi.hash().abs()
+
+proc execNimscript(nimsFile, actionName: string, options: Options, live = true): tuple[output: string, exitCode: int] =
+  var
+    cmd = ("nim e --hints:off --verbosity:0 " & nimsFile.quoteShell & " " & actionName).strip()
+
+  if live:
+    result.exitCode = execCmd(cmd)
+    let
+      outFile = nimsFile & ".out"
+    if outFile.fileExists():
+      result.output = outFile.readFile()
+      discard outFile.tryRemoveFile()
+  else:
+    result = execCmdEx(cmd, options = {poUsePath})
+
+proc setupNimscript*(scriptName: string, options: Options): tuple[nimsFile, iniFile: string] =
```

Comment:
Nit: This function's name is misleading, it doesn't _just_ setup the Nimscript, it also executes under certain circumstances. This should be reflected in the name.

I know you're caching the result, but it still feels wrong to always execute `pringPkgInfo` before running a task. I would put this logic in a separate procedure.

Reply:
Generating the nims file and saving the ini file is part of the setup. setupNimscript is kind of like setupVM in that sense. The execution is required, without running `nim e printPkgInfo`, nimble won't have package info. That's why it is bundled into the setup call.

Reply:
setupVM doesn't retrieve the package info though. Package info is only required when Nimble asks for it, when you're executing a task it isn't needed so why read it?

I would prefer to see this organised like this:

* executeNimscript(scriptName, action: string, options: Options): ExecutionResult[string]
* retrievePackageInfo(scriptName: string, options: Options): ExecutionResult[string]
  * This is where you call ``executeNimscript(scriptName, "printPkgInfo", options)`` and have all the .ini reading logic.

Does that make sense?

Reply:
I've cleaned this up - setupNimscript() only creates the nims and nimscriptapi.nim files and is not invoked outside the file. getIniFile() is the only proc exposed (apart from execHook and execTask) and creates the ini file and returns the path.